File: challenge-318/lubos-kolouch/python/ch-1.py
import unittest


def find_groups(s: str) -> list[str]:
    """
    Find all groups of three or more consecutive identical letters in a string.

    Args:
        s (str): The input string of lowercase letters.

    Returns:
        List[str]: A list of groups found in the string. Returns an empty list if no groups are found.
    """
    groups = []
    i = 0
    while i < len(s):
        count = 1
        while i + 1 < len(s) and s[i] == s[i + 1]:
            count += 1
            i += 1
        if count >= 3:
            groups.append(s[i] * count)
        i += 1
    return groups


class TestFindGroups(unittest.TestCase):

    def test_example_1(self):
        self.assertEqual(find_groups("abccccd"), ["cccc"])

    def test_example_2(self):
        self.assertEqual(find_groups("aaabcddddeefff"), ["aaa", "dddd", "fff"])

    def test_example_3(self):
        self.assertEqual(find_groups("abcdd"), [])

    def test_no_groups(self):
        self.assertEqual(find_groups("abcdef"), [])

    def test_single_group(self):
        self.assertEqual(find_groups("aaa"), ["aaa"])

    def test_multiple_groups(self):
        self.assertEqual(find_groups("aaabbbccc"), ["aaa", "bbb", "ccc"])


if __name__ == "__main__":
    unittest.main()

File: challenge-318/lubos-kolouch/perl/ch-1.pl
use strict;
use warnings;
use Test::More;

# Higher-order function to find groups
sub find_groups {
    my ($str) = @_;
    my @groups;

    # Use regex to find groups of 3 or more consecutive letters
    while ( $str =~ /([a-z])\1{2,}/g ) {
        push @groups, $&;
    }

    return @groups;
}

# Main script
sub group_position {
    my ($str) = @_;
    my @groups = find_groups($str);
    return @groups ? @groups : ('');
}

# Unit tests
sub run_tests {
    is_deeply( [ group_position("abccccd") ],        ["cccc"],                 'Test case 1' );
    is_deeply( [ group_position("aaabcddddeefff") ], [ "aaa", "dddd", "fff" ], 'Test case 2' );
    is_deeply( [ group_position("abcdd") ],          [""],                     'Test case 3' );
    is_deeply( [ group_position("aabbcc") ],         [""],                     'Test case 4 (no groups)' );
    is_deeply( [ group_position("aaa") ],            ["aaa"],                  'Test case 5 (single group)' );
    is_deeply( [ group_position("") ],               [""],                     'Test case 6 (empty string)' );
    done_testing();
}

# Run tests
run_tests();
